### ISSUE TITLE
Information about a Union type is lost when assigning the corresponding to a variable

### ISSUE THREAD

**Issue by Dobiasd** [python/mypy#10618]
The following code

```python3
from typing import Union, List

a = [42, 'hello']
b: List[Union[int, str]] = a
```

results in the error

`error: Incompatible types in assignment (expression has type "List[object]", variable has type "List[Union[int, str]]")`

when checked with `mypy`.

Assigning directly, however, is fine:

```python3
from typing import Union, List

b: List[Union[int, str]] = [42, 'hello']
```

`Success: no issues found in 1 source file`

So is using `float` instead of `str`:

```python3
from typing import Union, List

a = [42, 23.5]
b: List[Union[int, float]] = a
```

`Success: no issues found in 1 source file`

Here is a full `Dockerfile` example (`docker build .`) to reproduce the problem:

```Dockerfile
FROM python:3.9.5

RUN pip install mypy==0.901

RUN echo "\

from typing import Union, List\n\
\n\
a = [42, 'hello']\n\
b: List[Union[int, str]] = a\n\
" > main.py

RUN mypy main.py
```

**Comment by JelleZijlstra** [comment]
This is expected behavior due to mypy's treatment of type context.

**Comment by Dobiasd** [comment]
Ah, ok, thank you. :+1:
